### PR TITLE
[benchmark] Add new benchmark for floating-point conversion

### PR DESCRIPTION
**Pull request by xwu**
This PR adds a new benchmark for floating-point conversion.

Eventually, it will test more combinations, but for now, it looks to quantify the difference between concrete conversions from `Double` to `Double` and generic conversions from `Double` to `Double` in anticipation of upcoming PRs.

### CODE REVIEW

**Comment by xwu** [comment]
@swift-ci Please benchmark

File: benchmark/single-source/FloatingPointConversion.swift [hunk 1]
```diff
@@ -0,0 +1,66 @@
+//===--- FloatingPointConversion.swift ------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
```
Comment (benrimmington) [inline benchmark/single-source/FloatingPointConversion.swift:5]:
```suggestion
// Copyright (c) 2020 Apple Inc. and the Swift project authors
```

Comment (xwu) [reply benchmark/single-source/FloatingPointConversion.swift:5]: Yeah, we need to update the benchmark generator template.

**Comment by xwu** [comment]
@swift-ci smoke test

**Comment by xwu** [comment]
@swift-ci smoke test macOS platform

**Comment by xwu** [comment]
@troughton That's a good thought. For a thorough set of tests, though, we'll need to define mock types here, so it'll be a very long file even just with floating-point types, so I'd rather leave integer benchmarks to a separate file.

**Comment by xwu** [comment]
@swift-ci smoke test linux platform

**Comment by varungandhi-apple** [comment]
This benchmark (or at least, parts of it) should not be built for armv7. https://ci.swift.org/job/swift-PR-osx/23595/consoleText

```
FAILED: benchmark/Onone-armv7-apple-ios8.0/FloatingPointConversion.o 

/Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/swift/benchmark/single-source/FloatingPointConversion.swift:64:17: error: 'Float80' is unavailable: Float80 is not available on target platform.
  init(_ value: Float80) { self.init(_Value(value)) }
                ^~~~~~~
Swift.Float80:2:23: note: 'Float80' has been explicitly marked unavailable here
@frozen public struct Float80 {
                      ^
```

**Comment by eeckstein** [comment]
@xwu for optimizations where we expect the generated code to be trivial it is better to add a lit test than a micro benchmark.
In this case we expect the compiler to eliminate everything for those Float->Float conversions.
A micro-benchmark is not really the best option for validating this: if the code generation regresses it can still be missed if the change is not that big. Also, benchmarks are not mandatory to run in PR tests.

So, I looked at https://github.com/apple/swift/pull/33803 (nice work, BTW!) and found that the compiler still produces unnecessary code.
I fixed that deficiency in https://github.com/apple/swift/pull/33839.
I also added a lit test for testing the floating point conversion (test/SILOptimizer/floating_point_conversion.swift), which is a better test for your optimization than these benchmarks.

**Comment by xwu** [comment]
@eeckstein That's fantastic; thanks for your follow-up PR in #33839!

I have another optimization and refactoring in #33826 that I hope and expect will pass this new lit test as well. It emits code that's 3x faster despite not having the benefit of your changes in #33839, which I expected to be possible given the disparity between concrete performance and generic performance after landing #33803--but now I'm curious as to how the improvement was possible without tweaking the compiler.

This particular PR here was just a start to the benchmarks; I created mock wrapper types in #33821 to test further optimizations with custom floating-point types, which I did not expect would result in code that could be optimized to the point of being trivial. It's looking like all of it might optimize pretty well, though, so I do wonder as to the future of these benchmarks.

**Comment by eeckstein** [comment]
@xwu Sounds good. Feel free to add more tests in floating_point_conversion.swift.
The ConvertFloatingPoint.MockFloat64ToDouble benchmark makes sense.

I still recommend to remove the ConvertFloatingPoint.GenericDoubleToDouble/ConcreteDoubleToDouble benchmarks. They don't add much value, because the lit test already checks that the optimization is optimal.
In general, we should try to keep our set of benchmarks as small as possible to reduce the benchmark test time.
Alternatively you can add the ".skip" tag, so that the benchmarks are still available, but are excluded from the standard benchmark run.

**Pull request merged**